moabs: add more options

lijinbio opened the pull request:
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

bgruening commented on tools/moabs/moabs.xml:
```diff
@@ -147,44 +187,98 @@
-            <param argument="-n" type="boolean" truevalue="1" falsevalue="0" checked="true" label="Mapping to four strands?" help="Yes: map SE or PE reads to all 4 strands, i.e. ++, +-, -+, --; No: only map to 2 forward strands, i.e. BSW(++) and BSC(-+)"/>
-            <param argument="-r" type="select" label="How to report repeat hits" help="0=none(unique hit/pair); 1=random one; 2=all(slow)">
+            <param argument="-w" type="integer" value="1000" min="0" max="1000" label="Maximum number of equal best hits to count" help="Maximum number of equal best hits to count. When multiple mapping occurs for a read, it should control the number of records to report. Default: 1000."/>
+            <param argument="-D" type="text" label="Restriction enzyme digestion sites for RRBS mode" help="For RRBS data, this option activates RRBS mapping mode and set restriction enzyme digestion sites. Digestion position marked by '-', example: -D C-CGG for MspI digestion.  default: none (WGBS mode)."/>
```
Can you please add a validator and restrict the user input.

lijinbio replied:
Yes, a validator is added for user input.

bgruening commented on tools/moabs/moabs.xml:
```diff
@@ -147,44 +187,98 @@
+                <option value="64">Illumina</option>
+            </param>
+            <param argument="-f" type="integer" value="5" min="0" label="Maximum number of Ns in a read to filter out" help="To filter out low-quality reads containing >n Ns. Default=5."/>
+            <param argument="-A" type="text" label="3' adapter sequence to trim" help="To trim 3' adapter sequence. Default: none (no trim)."/>
```
Please add validator.

lijinbio replied:
Yes, a validator is added for the text inputs.

bgruening commented on tools/moabs/moabs.xml:
```diff
@@ -147,44 +187,98 @@
-            <option value="2"> comp.g1.vs.g2.txt </option>
-            <option value="3"> BAM files </option>
-            <option value="4"> Methylation calling BED files </option>
+            <option value="1" selected="true"> dmc_M2_g1.G.bed_vs_g2.G.bed.txt </option>
```
can we remove the first space?

lijinbio replied:
Yes, the first and last spaces are removed.

lijinbio commented:
Hi @bgruening, thank you for your comments. I will address them and update the PR accordingly.

lijinbio commented:
> @lijinbio this is awesome. Can you please bump the version number of the tool.

As for the version number of the tool, may I set 1.3.4.6.1 as the version id in the header? although the requirement of the MOABS version is not changed, still depending on MOABS 1.3.4.6.

bgruening commented:
@lijinbio please see here: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions

lijinbio commented:
@bgruening Hi Björn, I have addressed your comments, and the commit has passed the lint checks. Can you help review my PR again? Thank you.

bgruening commented:
@lijinbio sorry, unfortunaltey, the test are failling with:

Files /tmp/tmpyquix209SE_g1_r1.bam=96198b but /tmp/tmp8n3r_2vfSE_g1_r1.bam=111199b - compare by size (delta=10000) failed


I know its green sorry, we changed the testing recently and still need to iron out something.

The error indicates that your BAM files do have changed. BAM files by the way can be compared with normal diff, as Galaxy converts them to SAM and then compares the SAM files. So please remove the sim_size in your BAM test outputs.

lijinbio commented:
> @lijinbio sorry, unfortunaltey, the test are failling with:
> 
> Files /tmp/tmpyquix209SE_g1_r1.bam=96198b but /tmp/tmp8n3r_2vfSE_g1_r1.bam=111199b - compare by size (delta=10000) failed
> 
> I know its green sorry, we changed the testing recently and still need to iron out something.
> 
> The error indicates that your BAM files do have changed. BAM files by the way can be compared with normal diff, as Galaxy converts them to SAM and then compares the SAM files. So please remove the sim_size in your BAM test outputs.

@bgruening Hi Björn, yes, I realized I have added some alignment options to BSMAP, and the generated BAM may change w.r.t. the command line in the header. May I ask how could I run the test locally so that I can generate the exact BAM files by mimicking the web interface use? Thank you.

At the same time, I will change the test command as below. Do they follow the convention to test a BAM file?

```
            <output_collection name="output_collection_bam" count="4">
                <element name="g1_r1.bam" file="SE_g1_r1.bam" ftype="bam"/>
                <element name="g1_r2.bam" file="SE_g1_r2.bam" ftype="bam"/>
                <element name="g2_r1.bam" file="SE_g2_r1.bam" ftype="bam"/>
                <element name="g2_r2.bam" file="SE_g2_r2.bam" ftype="bam"/>
            </output_collection>
```

lijinbio commented:
@bgruening For example, the BAM header looks like below.

```
@HD	VN:1.0	SO:coordinate
@SQ	SN:chr11	LN:30000
@PG	ID:BSMAP	VN:2.87	CL:"bsmap -a /tmp/tmpmjC4hb/files/2/2/a/dataset_22a0b1fd-69cf-4fbe-a4bd-f31d0aa36ba9.dat -o g1_r1.bam -n 1 -d /tmp/tmpmjC4hb/files/f/e/8/dataset_fe8e1e6e-fc58-4ee5-8b8a-8663fb954d17.dat -r 0 -R -p 1"
```

Yet it will definitely change when running in the CI check. May I ask how could I make sure my input BAM will match the test BAMs? Thank you.

lijinbio commented:
> @lijinbio sorry, unfortunaltey, the test are failling with:
> 
> Files /tmp/tmpyquix209SE_g1_r1.bam=96198b but /tmp/tmp8n3r_2vfSE_g1_r1.bam=111199b - compare by size (delta=10000) failed
> 
> I know its green sorry, we changed the testing recently and still need to iron out something.
> 
> The error indicates that your BAM files do have changed. BAM files by the way can be compared with normal diff, as Galaxy converts them to SAM and then compares the SAM files. So please remove the sim_size in your BAM test outputs.

@bgruening Hi Björn, I have removed the sim_size but used lines_diff for the BAM test. Can you help review my PR again? Thank you.

bgruening commented:
@lijinbio thanks a lot! This looks great!

lijinbio commented:
> @lijinbio thanks a lot! This looks great!

@bgruening Hi Björn, many thanks for reviewing and merging my PR.

I found that the interface has not been upgraded yet at https://usegalaxy.eu/root?tool_id=moabs, may I ask if I need to submit a PR to change the revision number in `tools_iuc.yaml.lock` in the usegalaxy-eu-tools repository? so that the interface will update accordingly. Thank you for your help.

lijinbio commented:
> > @lijinbio thanks a lot! This looks great!
> 
> @bgruening Hi Björn, many thanks for reviewing and merging my PR.
> 
> I found that the interface has not been upgraded yet at https://usegalaxy.eu/root?tool_id=moabs, may I ask if I need to submit a PR to change the revision number in `tools_iuc.yaml.lock` in the usegalaxy-eu-tools repository? so that the interface will update accordingly. Thank you for your help.

@bgruening Hi Björn, I have submitted a PR to add the revision of MOABS at UseGalaxy.eu (
https://github.com/usegalaxy-eu/usegalaxy-eu-tools/pull/291), can you help review my PR there? Thank you.

bgruening commented:
Awesome, thanks!
